File: october_2024/potd_09_10_2024.java


// User function Template for Java

/*class Node

class Node
{
   int data;
   Node right,down;

   Node(int data){
       this.data = data;
       right = null;
       down = null;
   }
}
*/
/*Function should return the head of the 2D LL.*/
class Solution {
    static Node head;
    static Node construct(int arr[][]) {
        // Add your code here.
        int n=arr.length;
        
        if(n==1)
        {
            Node newnode =new Node(arr[0][0]); 
            head=newnode;
            return head;
        }
        
        Node newnode =new Node(arr[0][0]); 
        head=newnode;
        
        Node currnode=newnode;
        Node firstnode=head;
        Node secondnode=head;
        
        for(int i=0;i<n;i++)
        {
            
            if(i>0)
            {
                
                Node newnode3=new Node(arr[i][0]);
                firstnode.down=newnode3;
                
                currnode=newnode3;
                firstnode=firstnode.down;
               
            }
            
    
            for(int j=1;j<n;j++)
            {
                Node newnode2=new Node(arr[i][j]);
                currnode.right=newnode2;
                currnode=currnode.right;
            }
            
        }
        
        firstnode = head;
        while (firstnode.down != null) 
        {
            Node row1 = firstnode;
            Node row2 = firstnode.down;

            while (row1 != null && row2 != null) 
            {
                row1.down = row2;
                row1 = row1.right;
                row2 = row2.right;
            }

            firstnode = firstnode.down;
        }
        
        
        return head;
    }
}

File: october_2024/potd_09_10_2024.cpp
#include <bits/stdc++.h>
using namespace std;

// structure of the node of the linked list is as
struct Node
{
    int data;
    Node *right, *down;

    Node(int x)
    {
        data = x;
        right = NULL;
        down = NULL;
    }
};

// function must return the pointer to the first element of the in linked list i.e. that
// should be the element at arr[0][0]
class Solution
{
public:
    Node *constructLinkedMatrix(vector<vector<int>> &mat)
    {
        if (mat.empty() || mat[0].empty())
        {
            return nullptr;
        }

        int rows = mat.size();
        int cols = mat[0].size();

        Node ***node = new Node **[rows];
        for (int i = 0; i < rows; i++)
        {
            node[i] = new Node *[cols];
        }

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                node[i][j] = new Node(mat[i][j]);
            }
        }

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                if (j < cols - 1)
                {
                    node[i][j]->right = node[i][j + 1];
                }
                if (i < rows - 1)
                {
                    node[i][j]->down = node[i + 1][j];
                }
            }
        }

        return node[0][0];
    }
};